src/solc_0.8/bundleSandSale/PolygonBundleSandSale.sol

```diff
@@ -0,0 +1,281 @@
+import "../asset/ERC1155ERC721.sol";
+
+/// @title PolygonBundleSandSale contract.
+/// @notice This contract receives bundles of: NFTs (ERC1155) + Sand.
+/// @notice Then those bundles are sold to users. Users can pay BaseCoin (Ethers) or Dais for the bundles.
+contract PolygonBundleSandSale is WithAdmin, IERC1155TokenReceiver {
+    bytes4 private constant ERC1155_RECEIVED = 0xf23a6e61;
+    bytes4 private constant ERC1155_BATCH_RECEIVED = 0xbc197c81;
+
+    event BundleSale(
+        uint256 indexed saleId,
+        uint256[] ids,
+        uint256[] amounts,
+        uint256 sandAmount,
+        uint256 priceUSD,
+        uint256 numPacks
+    );
+
+    event BundleSold(
+        uint256 indexed saleId,
+        address indexed buyer,
+        uint256 numPacks,
+        address token,
+        uint256 tokenAmount
+    );
+
+    Medianizer private _medianizer;
+    IERC20 private _dai;
+    IERC20 private _sand;
+    ERC1155ERC721 private _asset;
+
+    address payable private _receivingWallet;
+
+    struct Sale {
+        uint256[] ids; // NFT ids
+        uint256[] amounts; // Amount of sand
+        uint256 sandAmount; //
+        uint256 priceUSD; // Price in USD
+        uint256 numPacksLeft;
+    }
+
+    Sale[] private sales;
+
+    constructor(
+        IERC20 sandTokenContractAddress,
+        ERC1155ERC721 assetTokenContractAddress,
+        Medianizer medianizerContractAddress,
+        IERC20 daiTokenContractAddress,
+        address admin,
+        address payable receivingWallet
+    ) {
+        require(receivingWallet != address(0), "need a wallet to receive funds");
+        _medianizer = medianizerContractAddress;
+        _sand = sandTokenContractAddress;
+        _asset = assetTokenContractAddress;
+        _dai = daiTokenContractAddress;
+        _admin = admin;
+        _receivingWallet = receivingWallet;
+    }
+
+    /// @notice set the wallet receiving the proceeds
+    /// @param newWallet address of the new receiving wallet
+    function setReceivingWallet(address payable newWallet) external onlyAdmin {
+        require(newWallet != address(0), "receiving wallet cannot be zero address");
+        _receivingWallet = newWallet;
+    }
+
+    /**
+     * @notice Buys Sand Bundle with Ether
+     * @param saleId id of the bundle
+     * @param numPacks the amount of packs to buy
+     * @param to The address that will receive the SAND
+     */
+    function buyBundleWithEther(
+        uint256 saleId,
+        uint256 numPacks,
+        address to
+    ) external payable {
+        require(saleId > 0, "invalid saleId");
+        uint256 saleIndex = saleId - 1;
+        uint256 numPacksLeft = sales[saleIndex].numPacksLeft;
+        require(numPacksLeft >= numPacks, "not enough packs on sale");
+        sales[saleIndex].numPacksLeft = numPacksLeft - numPacks;
+
+        uint256 usdRequired = numPacks * sales[saleIndex].priceUSD;
+        uint256 ethRequired = getEtherAmountWithUSD(usdRequired);
```

Comment:
I think that getEtherAmountWithUSD must do a ceil not a floor ? 
So you always pay enough ethers

Reply:
This makes sense to me. This way we have a better UX also, avoiding problems with the transactions.

Reply:
I was confused about the units, if the u$s price is in u$s * 10e18 we can avoid the extra calculation (the error is just 1wei). For the UI this kind of things are always an issue.
Added a comment the code and a small test to show the consecuences.